source/request.js

```diff
@@ -4,8 +4,37 @@ const nodeFetch = require("node-fetch");
 
 let fetchMethod = nodeFetch;
 
-function request(url, options) {
-    return fetchMethod(url, options);
+/**
+ * return a URL encoded path, including encoding for special characters like #
+ * @param {String} uriPath that we're trying to encode
+ * @returns {String} encoded uri path
+ */
+function encodeURIPath(uriPath) {
+    const uriSections = uriPath.split("/");
+    for (let i = 0; i < uriSections.length; i += 1) {
+      uriSections[i] = encodeURIComponent(uriSections[i]);
+    }
+    return uriSections.join("/");
+}
+
+/**
+ * Parse the url in a way that won't crash if the URL is invalid (conventional methods will not accept characters like the #)
+ * @param {String} fullUrl that was sent to the request function
+ * @returns {Object} protocol, hostname, and path
+ */
+function splitInvalidURL(fullUrl){
+    const sections = fullUrl.split("/");
+    const url = {};
+    url.protocol = sections[0] + "//";
+    url.hostname = sections[2];
+    url.path = "/" + sections.slice(3, sections.length).join("/");
+    return url;
+}
+
+function request(fullUrl, options) {
```

Comment:
I wouldn't recommend changing this file to perform this work. For one, splitting the URL like this is a bit redundant and unsafe. Ultimately only the _path_ needs encoding ([here](https://github.com/perry-mitchell/webdav-client/blob/master/source/interface/putFile.js#L27)), which could be done in the interface files like [putFile.js](https://github.com/perry-mitchell/webdav-client/blob/master/source/interface/putFile.js).

Ideally there'd be some helper method to perform the encoding like `encodeRequestPath(path)` that you could use on all get/put/mkdir requests.

Reply:
I tried a bunch of different splitting methods, particularly `url = new URL(...)` however since something like 'https://webdav-server.com/webdav/file #1.json` has the `#` character, the path is only parsed up to that point (`#1.json` is removed). How would you go about parsing?

I'll look into moving the encoding around. I went this way to keep the code more DRY.